src/Microsoft.Azure.Batch.SoftwareEntitlement.Server/Controllers/SoftwareEntitlementsController.cs

```diff
@@ -53,63 +56,73 @@ public SoftwareEntitlementsController(Options options, ILogger logger)
-                entitlementRequest.Token,
-                _options.Audience,
-                _options.Issuer,
-                entitlementRequest.ApplicationId,
-                remoteAddress);
-            if (!verificationResult.HasValue)
-            {
-                foreach (var e in verificationResult.Errors)
+                    return StatusCode(400, error);
+                }
+
+                _logger.LogInformation(
+                    "Selected api-version is {ApiVersion}",
+                    apiVersion);
+
+                _logger.LogInformation(
+                    "Requesting entitlement for {Application}",
+                    entitlementRequest.ApplicationId);
+                _logger.LogDebug("Request token: {Token}", entitlementRequest.Token);
+
+                var remoteAddress = HttpContext.Connection.RemoteIpAddress;
+                _logger.LogDebug("Remote Address: {Address}", remoteAddress);
+
+                var verificationResult = _verifier.Verify(
+                    entitlementRequest.Token,
+                    _serverOptions.Audience,
+                    _serverOptions.Issuer,
+                    entitlementRequest.ApplicationId,
+                    remoteAddress);
+                if (!verificationResult.HasValue)
                 {
-                    _logger.LogError(e);
+                    foreach (var e in verificationResult.Errors)
+                    {
+                        _logger.LogError(e);
+                    }
+
+                    var error = new SoftwareEntitlementFailureResponse
+                    {
+                        Code = "EntitlementDenied",
+                        Message = new ErrorMessage($"Entitlement for {entitlementRequest.ApplicationId} was denied.")
+                    };
+
+                    return StatusCode(403, error);
                 }
 
-                var error = new SoftwareEntitlementFailureResponse
+                var entitlement = verificationResult.Value;
+                var response = new SoftwareEntitlementSuccessfulResponse
                 {
-                    Code = "EntitlementDenied",
-                    Message = new ErrorMessage($"Entitlement for {entitlementRequest.ApplicationId} was denied.")
+                    EntitlementId = entitlement.Identifier,
+                    VirtualMachineId = entitlement.VirtualMachineId
                 };
 
-                return StatusCode(403, error);
+                return Ok(response);
             }
-
-            var entitlement = verificationResult.Value;
-            var response = new SoftwareEntitlementSuccessfulResponse
+            finally
             {
-                EntitlementId = entitlement.Identifier,
-                VirtualMachineId = entitlement.VirtualMachineId
-            };
-
-            return Ok(response);
+                if (_serverOptions.Flags.HasFlag(ServerFlags.ExitAfterRequest))
+                {
```

Comment:
this feels like something we should be doing at the server level and not in the controller (e.g. adding something to the pipeline)

Reply:
That's a good idea - I've taken a note of it to do later.